package/samplers/bisect/_sampler.py

```diff
@@ -0,0 +1,199 @@
+    ) -> dict[str, Any]:
+        return {}
+
+    def _get_current_left_and_right(
+        self, study: Study, param_name: str
+    ) -> tuple[int | float, int | float]:
+        left_key = f"{PREFIX_LEFT}{param_name}"
+        right_key = f"{PREFIX_RIGHT}{param_name}"
+        system_attrs = study._storage.get_study_system_attrs(study._study_id)
+        left = system_attrs[left_key]
+        right = system_attrs[right_key]
+        assert isinstance(left, (int, float)) and isinstance(right, (int, float))
+        return left, right
+
+    def _set_left_and_right(
+        self,
+        study: Study,
+        param_name: str,
+        *,
+        left: int | float | None = None,
+        right: int | float | None = None,
+    ) -> None:
+        left_key = f"{PREFIX_LEFT}{param_name}"
+        right_key = f"{PREFIX_RIGHT}{param_name}"
+        if left is not None:
+            study._storage.set_study_system_attr(study._study_id, left_key, left)
+        if right is not None:
+            study._storage.set_study_system_attr(study._study_id, right_key, right)
+
+    def sample_independent(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        param_name: str,
+        param_distribution: BaseDistribution,
+    ) -> Any:
+        if isinstance(param_distribution, optuna.distributions.CategoricalDistribution):
+            raise ValueError("CategoricalDistribution is not supported.")
+        if param_distribution.log:
+            raise ValueError("Log scale is not supported.")
+
+        self._search_space[param_name] = param_distribution
+        step = param_distribution.step
+        is_discrete = step is not None
+        if trial.number == 0:
+            right = param_distribution.high + step if is_discrete else param_distribution.high
+            self._set_left_and_right(study, param_name, left=param_distribution.low, right=right)
+
+        left, right = self._get_current_left_and_right(study, param_name)
+        if not is_discrete:
+            mid = (left + right) / 2.0
+            return mid
+
+        possible_param_values = self._get_possible_param_values(param_distribution)
+        indices = np.arange(len(possible_param_values))
+        left_index = indices[np.isclose(possible_param_values, left)][0]
+        right_index = indices[np.isclose(possible_param_values, right)][0]
+        mid_index = (right_index + left_index) // 2
+        assert mid_index != len(possible_param_values) - 1, "The last element is for convenience."
+        return possible_param_values[mid_index].item()
+
+    def _get_possible_param_values(
+        self, param_distribution: FloatDistribution | IntDistribution
+    ) -> np.ndarray:
+        step = param_distribution.step
+        low = param_distribution.low
+        # The last element is padded to code the binary search routine cleaner.
+        high = param_distribution.high + step
+        assert step is not None
```

Comment:
How about moving this assertion here, since step is already used in the addition operation, which doesn’t allow `None`?
```suggestion
        assert step is not None
        low = param_distribution.low
        # The last element is padded to code the binary search routine cleaner.
        high = param_distribution.high + step
```